DevTools/SourceMaps/explainer.md

```diff
@@ -0,0 +1,245 @@
+
+## Goals
+
+ * Improve post-hoc debugging capabilities
+ * Preserve compatibility with existing tooling
+
+## Non-Goals
+
+Nothing specific at this time.
+
+## Use Cases
+
+*The following was taken from [an issue opened on the TypeScript repo](https://github.com/microsoft/TypeScript/issues/46695):*
+
+Suppose I receive the following information from my application:
+
+```
+TypeError: Cannot read properties of undefined (reading 'x')
+    at re.draw (bundle.js:1:16372)
+    at re.drawLayer (bundle.js:1:14170)
+    at be.draw (bundle.js:1:74592)
+    at Te.render (bundle.js:1:114230)
+    at Te.reflowCanvas (bundle.js:1:113897)
+    at HTMLDocument.anonymous (bundle.js:1:135849)
+```
+
+The source map will resolve the first stack frame to `draw(src, sprite, x, y, frame)`, and will correctly point out that the failure here is that the undefined value is actually `frame`. However, *that is not the name of the function*. The function's name is `draw`, which is a member of the `Render` class.
+
+If I simply apply the value of the `names` array to the function, decoding the stack trace is not particularly useful. It would look something like this:
+
+```
+TypeError: Cannot read properties of undefined (reading 'x')
+    at frame (file1.ts:302:7)
+    at draw (file1.ts:144:5)
+    at Render (file2.ts:95:5)
+    at drawArray (file3.ts:178:5)
+    at render (file3.ts:155:5)
+    at game (file4.ts:39:5)
+```
+
+Using the source map to navigate the source code by hand, I can reconstruct the original call stack:
+
+```
+TypeError: Cannot read properties of undefined (reading 'x')
+    at Render#draw (file1.ts:302:7)
+    at Render#drawLayer (file1.ts:144:5)
+    at GameObject#draw (file2.ts:95:5)
+    at Game#render (file3.ts:178:5)
+    at Game#reflowCanvas (file3.ts:155:5)
+    at [anonymous function passed to addEventListener] (file4.ts:39:5)
+```
+
+But doing this required that I dump the mappings and the source files from the source map and manually inspect the source files. In general, this requires that I use a library to parse the `mappings` field from Source Maps (because [as of Source Maps v3](https://sourcemaps.info/spec.html), this field is stored in a stateful way).
+
+## Proposed Solution
+
+We want to adopt the general approach taken by the `pasta-sourcemaps` ("Pretty (and) Accurate Stack Trace Analysis") library. However, instead of adding to the `names` field, we will add two fields.
+
+* Add two additional field to the source map: `scopes` and `scopeNames`
+* The `scopes` field should be a list of ranges and pointers into `scopeNames`
+* The ranges in the field should always point to **Original Source** locations.

```

Comment:
Untangling these from `names` is probably cleaner. Might be be worth speccing here the expectation of the sourcemap "apply" operation needed to combine sourcemaps? (considering that a sourcemap combine stage may be combining N earlier maps with 1 new map in the case of a transformation like inlining).

Reply:
Well, I was thinking about this, too. Since the original change, I pivoted that there's only one `scopes` field -- not an array of `scopes` fields -- but the bundler will need to process these either way. I'm working on that with webpack/webpack#14996.